src/Domain/Asset/ValueObject/UploadTarget.php

```
@@ -0,0 +1,121 @@
+        // Wrap IPv6 hosts in brackets when rebuilding the URL. Use the
+        // normalized host (without any surrounding brackets) to avoid
+        // double-bracketing when parse_url returned a bracketed host.
+        $hostForUrl = strpos($host, ':') !== false ? '[' . $host . ']' : $host;
+
+        $port = isset($parsedUrl['port']) ? ':' . $parsedUrl['port'] : '';
+        $path = $parsedUrl['path'] ?? '';
+        $query = isset($parsedUrl['query']) ? '?' . $parsedUrl['query'] : '';
+        $fragment = isset($parsedUrl['fragment']) ? '#' . $parsedUrl['fragment'] : '';
+
+        return $scheme . '://' . $auth . $hostForUrl . $port . $path . $query . $fragment;
+    }
```

PHPMD (2.15.0)

[warning] The method normalizeUrl() has a Cyclomatic Complexity of 14. The configured cyclomatic complexity threshold is 10. (undefined)

(CyclomaticComplexity)